docs: content for KeptnTaskDefinition ref and tasks guide (#1392)
keptn/lifecycle-toolkit: docs/content/en/docs/yaml-crd-ref/taskdefinition.md

description: Define tasks that can be run pre- or post-deployment
weight: 89
---

A `KeptnTaskDefinition` defines tasks
that are run by the Keptn Lifecycle Toolkit
as part of the pre- and post-deployment phases of a
[KeptnApp](./app.md) or

> [workload](../concepts/workloads/).

Reviewer: here we should refer to KeptnWorkload

Author: done

## Yaml Synopsis

```yaml
apiVersion: lifecycle.keptn.sh/v?alpha?
kind: KeptnTaskDefinition
metadata:
  name: <task-name>
spec:
  function:
    inline | httpRef | functionRef:

>     [parameters:
>       map:
>         textMessage: "This is my configuration"]
>     [secureParameters:
>       secret: slack-token]

Reviewer: Let's not use these linux optional brackets as it is used in linux man pages. It's not very common in kubernetes docs
```suggestion
    parameters:
      map:
        textMessage: "This is my configuration"
    secureParameters:
      secret: slack-token
```

      In this case, it calls other, existing `KeptnTaskDefinition`s
      for each type of test to be run,
      specifying each by the value of the `name` field.

  * **parameters** - An optional field to supply input parameters to a function.
    The Lifecycle Toolkit passes the values defined inside the `map` field
    as a JSON object.
    For example:

     ```yaml
       spec:
         parameters:
           map:
             textMessage: "This is my configuration"
     ```

     See
     [Parameterized functions](../implementing/tasks/#parameterized-functions)
     for more information.

  * **secureParameters** -- An optional field used to pass a Kubernetes secret.
    The `secret` value is the Kubernetes secret name
    that is mounted into the runtime and made available to functions
    using the `SECURE_DATA` environment variable.
    For example:

    ```yaml
    secureParameters:
      secret: slack-token
    ```

    Note that, currently, only one secret can be passed.

    See [Create secret text](../implementing/tasks/#create-secret-text)
    for details.

## Usage

A Task executes the TaskDefinition of a
[KeptnApp](app.md) or

> [workload](https://kubernetes.io/docs/concepts/workloads/).

Reviewer: We should refer to KeptnWorkload here

Author: done

and the revised code is picked up without additional changes.

## Examples

### Example 1: inline script

This example defines a full-fledged Deno script
within the `KeptnTaskDefinition` YAML file:

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha3
kind: KeptnTaskDefinition
metadata:
  name: hello-keptn-inline
spec:
  function:
    inline:
      code: |
        let text = Deno.env.get("DATA");
        let data;
        let name;
        data = JSON.parse(text);

        name = data.name
        console.log("Hello, " + name + " new");
```

### Example 2: httpRef script

This example fetches the Deno script from a remote webserver at runtime:

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha3
kind: KeptnTaskDefinition
metadata:
  name: hello-keptn-http
spec:
  function:
    httpRef:
      url: "https://www.example.com/yourscript.js"
```

For another example, see the
[sample-app](https://github.com/keptn-sandbox/lifecycle-toolkit-examples/blob/main/sample-app/version-1/app-pre-deploy.yaml).

See the
[sample-app/version-1](https://github.com/keptn-sandbox/lifecycle-toolkit-examples/blob/main/sample-app/version-1/app-pre-deploy.yaml)
PodtatoHead example for a more complete example.

> ### Example 3: functionRef

Reviewer: To the examples, I would maybe also add an example about referencing a configmap, which contains the function code

Author: Do you know where I can find such an example?

     ```

     See
     [Parameterized functions](../implementing/tasks/#parameterized-functions)
     for more information.

  * **secureParameters** -- An optional field used to pass a Kubernetes secret.
    The `secret` value is the Kubernetes secret name
    that is mounted into the runtime and made available to functions
    using the `SECURE_DATA` environment variable.
    For example:

    ```yaml
    secureParameters:
      secret: slack-token
    ```

    Note that, currently, only one secret can be passed.

    See [Create secret text](../implementing/tasks/#create-secret-text)
    for details.

## Usage

A Task executes the TaskDefinition of a
[KeptnApp](app.md) or [KeptnWorkload].
The execution is done by spawning a Kubernetes
[Job](https://kubernetes.io/docs/concepts/workloads/controllers/job/)
to handle a single Task.
In its state, it tracks the current status of this Kubernetes Job.

The `function` is coded in JavaScript
and executed in
[Deno](https://deno.com/runtime),
which is a lightweight runtime environment
that executes in your namespace.
Note that Deno has tighter restrictions
for permissions and importing data
so a script that works properly elsewhere

> may not function properly when run in Deno.

Reviewer 2: ```suggestion
may not function out of the box when run in Deno.
```

        let text = Deno.env.get("DATA");
        let data;
        let name;
        data = JSON.parse(text);

        name = data.name
        console.log("Hello, " + name + " new");
```

### Example 2: httpRef script

This example fetches the Deno script from a remote webserver at runtime:

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha3
kind: KeptnTaskDefinition
metadata:
  name: hello-keptn-http
spec:
  function:
    httpRef:
      url: "https://www.example.com/yourscript.js"
```

For another example, see the
[sample-app](https://github.com/keptn-sandbox/lifecycle-toolkit-examples/blob/main/sample-app/version-1/app-pre-deploy.yaml).

See the
[sample-app/version-1](https://github.com/keptn-sandbox/lifecycle-toolkit-examples/blob/main/sample-app/version-1/app-pre-deploy.yaml)
PodtatoHead example for a more complete example.

### Example 3: functionRef

This example calls another defined task,
illustrating how one `KeptnTaskDefinition` can build
on top of other `KeptnTaskDefinition`s.
In this case, it calls `slack-notification-dev`,
passing `parameters` and `secureParameters` to that other task:

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha3
kind: KeptnTaskDefinition
metadata:

>   name: slack-notification-dev

Reviewer 2: it would be nice if this example would use a task def from one of the two above, so that the user can see both tasks for clarity. So maybe we can substitute theslack notification dev with hello keptn example and pass a new name in the parameters

Author: Agreed!  Do we have an existing example that I could just use instead of this one from Andi?

And could we defer this fix to another PR?  I'd like to get this cleaned up and get a new, healthy fork...

Reviewer 2: sure we can merge as is and I can send you the suggested yaml after

spec:
  function:
    functionRef:
      name: slack-notification
    parameters:
      map:
        textMessage: "This is my configuration"
    secureParameters:
      secret: slack-token
```

> ### Example 4: ConfigMapRef

Reviewer 2: can we add the config map too here? you can use this example https://github.com/keptn/lifecycle-toolkit/blob/main/operator/config/samples/function_execution/taskdefinition_function_configmap.yaml

Author: Added

This example references a `ConfigMap` by the name of `dev-configmap`
that contains the code for the function to be executed.

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha3
kind: KeptnTaskDefinition
metadata:
  name: keptntaskdefinition-sample
spec:
  function:
    configMapRef:
      name: dev-configmap
```

### More examples

See the [operator/config/samples](https://github.com/keptn/lifecycle-toolkit/tree/main/operator/config/samples)
directory for more example `KeptnTaskDefinition` YAML files.
Separate examples are provided for each API version.
For example, the `lifecycle_v1alpha3_keptntaskdefinition` file
contains examples for the `v1alpha3` version of the lifecycle API group.

> 

Reviewer 2: You can probably also reference the function execution samples folder as well for task definition examples https://github.com/keptn/lifecycle-toolkit/tree/main/operator/config/samples/function_execution

Author: Done

      ```

    * **httpRef** - Specify a Deno script to be executed at runtime
      from the remote webserver that is specified.
      For example:

      ```yaml
      name: hello-keptn-http
        spec:
            function:
              httpRef:
                url: "https://www.example.com/yourscript.js"
      ```

    * **functionRef** -- Execute one or more `KeptnTaskDefinition` resources
      that have been defined.
      Populate this field with the value(s) of the `name` field
      for the `KeptnTaskDefinition`(s) to be called.
      This is commonly used to call a general function
      that is used in multiple places, possibly with different parameters.
      An example is:

       ```yaml
       spec:
         function:
           functionRef:
             name: slack-notification
       ```

      This can also be used to group a set of tasks
      into a single `KeptnTaskDefinition`,
      such as defining a `KeptnTaskDefinition` for testing.
      In this case, it calls other, existing `KeptnTaskDefinition`s
      for each type of test to be run,
      specifying each by the value of the `name` field.

>       

Reviewer 3: ```suggestion

```

Reviewer 3: this will fix the markdown lint issue

description: Define tasks that can be run pre- or post-deployment
weight: 89
---

A `KeptnTaskDefinition` defines tasks
that are run by the Keptn Lifecycle Toolkit
as part of the pre- and post-deployment phases of a

> [KeptnApp](./app.md).

Reviewer 4: ```suggestion
[KeptnApp](./app.md) or [KeptnWorkload](./workload.md).
```

Author: @thisthat We do not have a ref page for the KeptnWorkload yaml file because I think it is autogenerated and users do not need to populate it.  Do we want a KeptnWorkload page in the yaml-crd-ref directory or shall I link to [Keptn automatic app discovery](../implementing/integrate.md/#use-keptn-automatic-app-discovery) ?

Author: For the time being, I added a link to the "Workloads" page in "Concepts".    Probably not the most perfect solution for the long run but we need to revisit all sorts of structural issues in the docs after we get these Getting Started guides published.

## Yaml Synopsis

```yaml
apiVersion: lifecycle.keptn.sh/v?alpha?
kind: KeptnTaskDefinition
metadata:
  name: <task-name>
spec:
  function:

>     inline | httpRef | functionRef:
>     [parameters:
>       map:
>         textMessage: "This is my configuration"]
>     [secureParameters:
>       secret: slack-token]

Reviewer: ```suggestion
    httpRef:
      url: "https://www.example.com/yourscript.js"
    parameters:
      key: "value"
    secureParameters:
      secret: slack-token
```

Author: I think the Synopsis should use pseudo-code.  People can copy/paste the Example code.

* **metadata**
  * **name** -- Unique name of this task.
    Names must comply with the
    [Kubernetes Object Names and IDs](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names)
    specification.

* **spec**
  * **function** -- Code to be executed,
    expressed as a [Deno](https://deno.land/) script.
    Refer to [function runtime](https://github.com/keptn/lifecycle-toolkit/tree/main/functions-runtime)
    for more information about the runtime.

    The `function` can be defined as one of the following:

    * **inline** - Include the actual executable code to execute.
      This can be written as a full-fledged Deno script
      that is included in this file.
      For example:

      ```yaml
      function:
        inline:
          code: |
            console.log("Deployment Task has been executed");
      ```

    * **httpRef** - Specify a Deno script to be executed at runtime
      from the remote webserver that is specified.
      For example:

      ```yaml
      name: hello-keptn-http
        spec:
            function:
              httpRef:
                url: "https://www.example.com/yourscript.js"
      ```

>       Note that the file referenced is actually JavaScript,

Reviewer 4: I think we can leave out this paragraph to avoid confusion, since Deno has tighter restrictions when it comes to permissions or importing other libraries. So it might be the case that a script working in e.g. nodeJS will not function properly in Deno

Author: Ugh!  We have the example that is calling JavaScript and I notice that Andi and others are referring to this as JavaScript.  And we don't say anything about Deno having tighter restrictions.  Let me do a bit of a rewrite here.

Author: I removed this sentence and added a paragraph about Deno execution in the "Usage" section.

      which is essentially the same as a Deno script.

    * **functionRef** -- Execute another `KeptnTaskDefinition` that has been defined.
      Populate this field with the value of the `name` field
      for the `KeptnTaskDefinition` to be called.
      This is commonly used to call a general function
      that is used in multiple places, possibly with different parameters.
      An example is:

       ```yaml
       spec:
         function:
           functionRef:
             name: slack-notification
       ```

      This can also be used to group a set of tasks
      into a single `KeptnTaskDefinition`,
      such as defining a `KeptnTaskDefinition` for testing.
      In this case, it other, existing `KeptnTaskDefinition`s
      for each type of test to be run,
      specifying each by the value of the `name` field.

  * **parameters** - An optional field to supply input parameters to a function.

>     The Lifecycle Toolkit passes the values defined inside the `map` field

Reviewer: ```suggestion
    The Lifecycle Toolkit passes these values
```

    as a JSON object.
    For example:

     ```yaml
       spec:

>          parameters:
>            map:
>              textMessage: "This is my configuration"

Reviewer: ```suggestion
         parameters:
           key: "value"
```

     ```

     The JSON object can be read
     through the `DATA` environment variable using `Deno.env.get("DATA");`.

     Multi-level maps are not supported at this time.

     Currently only one secret can be passed.
     The secret must have a `key` called `SECURE_DATA`.
     It can be accessed via the environment variable `Deno.env.get("SECURE_DATA")`.

>      See [Context](#context) for details.

Author: I agree.  It's done.

  * **secureParameters** -- An optional field used to pass a Kubernetes secret.
    The `secret` value is the Kubernetes secret name
    that is mounted into the runtime and made available to functions
    using the `SECURE_DATA` environment variable.
    For example:

    ```yaml
    secureParameters:
      secret: slack-token
    ```

    See [Create secret text](#create-secret-text) for details.

## Usage

A Task is responsible for executing the TaskDefinition of a

> [workload](https://kubernetes.io/docs/concepts/workloads/)..

Reviewer: Here we can say, that workload or application, as on both levels, the principles are the same

The execution is done by spawning a Kubernetes
[Job](https://kubernetes.io/docs/concepts/workloads/controllers/job/)
to handle a single Task.
In its state, it tracks the current status of this Kubernetes Job.

### Context

A Kubernetes context is a set of access parameters
that contains a Kubernetes cluster, a user, and a namespace.
For more information, see
[Configure Access to Multiple Clusters](https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/).

A context environment variable is available via `Deno.env.get("CONTEXT")`.
It can be used like this:

```javascript
let context = Deno.env.get("CONTEXT");

> if (contextdata.objectType == "Application") {

Reviewer: ```suggestion
if (context.objectType == "Application") {
```

    let application_name = contextdata.appName;
    let application_version = contextdata.appVersion;
}

> if (contextdata.objectType == "Workload") {

Reviewer: ```suggestion
if (context.objectType == "Workload") {
```

    let application_name = contextdata.appName;
    let workload_name = contextdata.workloadName;
    let workload_version = contextdata.workloadVersion;
}
```

### Create secret text

```yaml

> # kubectl create secret generic my-secret --from-literal=SECURE_DATA=foo

Reviewer: I find this a little bit misleading, as it seems that it's part of yaml. We can maybe add a paragraph like:

To create a secret used in the KeptnTask execute this command:
kubectl create secret generic my-secret --from-literal=SECURE_DATA=foo

Or something in this manner.

Users might find it confusing seeing a kubectl command together with a yaml manifest

Author: Good stuff!  I added your lead-in to the command and also moved this section to the guide chapter, along with the info about context.

apiVersion: lifecycle.keptn.sh/v1alpha3
kind: KeptnTaskDefinition
metadata:
  name: dummy-task
  namespace: "default"
spec:
  function:
    secureParameters:
      secret: my-secret
    inline:
      code: |
        let secret_text = Deno.env.get("SECURE_DATA");
        // secret_text = "foo"
```

This methodology supports multiple variables
by creating a Kubernetes secret with a JSON string:

```yaml
# kubectl create secret generic my-secret \

> # --from-literal=SECURE_DATA="{\"foo\": \"bar\", \"foo2\": \"bar2\"}"

Author: I agree.  I added your lead-in to the command and also moved this whole section to the guide  page in "Implementing...".
